File: J9Ch/src/J_6_Linked_List_Array/Related_8/python_cpp/_621_Maximum_Subarray_V_Hard.py
# /** 621  Maximum Subarray V


class Solution:
    # @param {int[]} nums an array of integers
    # @param {int} k1 an integer
    # @param {int} k2 an integer
    # @return {int} the largest sum
    def maxSubarray5(self, nums, k1, k2):
        # Write your code here
        n = len(nums)
        if n < k1:
            return 0

        import sys
        result = -sys.maxint

        sum = [0 for _ in xrange(n + 1)]

        from collections import deque
        queue = deque()

        for i in xrange(1, n + 1):
            sum[i] = sum[i - 1] + nums[i - 1]
            if len(queue) and queue[0] < i- k2:
                queue.popleft()

            if i >= k1:
                while len(queue) and sum[queue[-1]] > sum[i - k1]:
                    queue.pop()
                queue.append(i - k1)

            # [i - k2, i - k1]
            if len(queue) and sum[i] - sum[queue[0]] > result:
                result = max(result, sum[i] - sum[queue[0]])

        return result

File: JS9Ch/src/J_6_Linked_List_Array/Related_8/python_cpp/_621_Maximum_Subarray_V_Hard.cpp
/** 621  Maximum Subarray V */


class Solution {
public:
    /**
     * @param nums an array of integers
     * @param k1 an integer
     * @param k2 an integer
     * @return the largest sum
     */
    int maxSubarray5(vector<int>& nums, int k1, int k2) {
        // Write your code here
        int n = nums.size();
        if (n < k1)
            return 0;

        int result = INT_MIN;

        vector<int> sum(n + 1);
        sum[0] = 0;

        // 6 [2, 4]
        deque<int> queue;
        for (int i = 1; i <= n; ++i) {
            sum[i] = sum[i - 1] + nums[i - 1];
            if (!queue.empty() && queue.front() < i- k2) {
                queue.pop_front();
            }
            if (i >= k1) {
                while (!queue.empty() && sum[queue.back()] > sum[i - k1]) {
                    queue.pop_back();
                }
                queue.push_back(i - k1);
            }

            // [i - k2, i - k1]
            if (!queue.empty() && sum[i] - sum[queue.front()] > result) {
                result = max(result, sum[i] - sum[queue.front()]);
            }
        }
        return result;
    }
};